### PR TITLE
add maxBytesInMemory and maxClientResponseBytes to SamplerConfig

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
Adds additional limits `maxBytesInMemory` and `maxClientResponseBytes` to the Druid sampler API config, to augment existing options of `numRows` and `timeoutMs`. While these two configs are quite similar, `maxBytesInMemory` is oriented around controlling how big the `IncrementalIndex` which performs the sampling will be, while `maxClientResponseBytes` is an attempt to allow clients to specify the maximum size of response they would prefer to handle, which might vary significantly from `maxBytesInMemory`.

`maxClientResponseBytes` is done in a rather sad way of serializing (and throwing away) the sampler rows and counting the bytes. In the future we might want to rework the sampler API to be a streaming API and do the counting of this stuff and enforcement of limits a bit more naturally.

This PR has:
- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/SamplerConfig.java [hunk 1]
```diff
@@ -22,34 +22,53 @@
...
+
+  private final long maxClientResponseBytes;
+
   @JsonCreator
   public SamplerConfig(
-      @JsonProperty("numRows") Integer numRows,
-      @JsonProperty("timeoutMs") Integer timeoutMs
+      @JsonProperty("numRows") @Nullable Integer numRows,
+      @JsonProperty("timeoutMs") @Nullable Integer timeoutMs,
+      @JsonProperty("maxBytesInMemory") @Nullable Long maxBytesInMemory,
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/SamplerConfig.java:50]:
```suggestion
      @JsonProperty("maxBytesInMemory") @Nullable HumanReadableBytes maxBytesInMemory,
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/SamplerConfig.java [hunk 2]
```diff
@@ -22,34 +22,53 @@
...
+      @JsonProperty("maxClientResponseBytes") @Nullable Long maxClientResponseBytes
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/SamplerConfig.java:51]:
```suggestion
      @JsonProperty("maxClientResponseBytes") @Nullable HumanReadableBytes maxClientResponseBytes
```

File: server/src/main/java/org/apache/druid/server/http/SegmentListerResource.java [hunk 3]
```diff
@@ -253,6 +253,7 @@ public void applyDataSegmentChangeRequests(
 
     final AsyncContext asyncContext = req.startAsync();
 
+
     asyncContext.addListener(
```
Comment (FrankChen021) [inline server/src/main/java/org/apache/druid/server/http/SegmentListerResource.java:257]: accidental change?

**Review by FrankChen021 — APPROVED** [review]
LGTM with minor comments

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/SamplerConfig.java (hunk 2 above)
Comment (clintropolis) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/SamplerConfig.java:51]: yeah, that seems reasonable, forgot about that thing 👍

**Pull request merged**